File: java/src/main/java/com/penapereira/patterns/interpreter/Context.java
package com.penapereira.patterns.interpreter;

import java.util.HashMap;
import java.util.Map;
import java.util.NoSuchElementException;

/** Information global to the interpreter: the variable bindings. */
public final class Context {

    private final Map<String, Integer> variables = new HashMap<>();

    public Context assign(String name, int value) {
        variables.put(name, value);
        return this;
    }

    public int lookup(String name) {
        Integer value = variables.get(name);
        if (value == null) {
            throw new NoSuchElementException("undefined variable: " + name);
        }
        return value;
    }
}

File: python/src/patterns/interpreter/context.py
class Context:
    """Information global to the interpreter: the variable bindings."""

    def __init__(self) -> None:
        self._variables: dict[str, int] = {}

    def assign(self, name: str, value: int) -> "Context":
        self._variables[name] = value
        return self

    def lookup(self, name: str) -> int:
        if name not in self._variables:
            raise KeyError(f"undefined variable: {name}")
        return self._variables[name]
